lib/graphlient/client.rb

```diff
@@ -1,37 +1,40 @@
+require 'graphql/client'
+require 'graphlient/adapters/faraday_adapter'
+
 module Graphlient
   class Client
-    attr_reader :uri
-    attr_reader :options
+    attr_accessor :uri, :options
 
     def initialize(url, options = {})
-      @uri = URI(url)
+      @url = url
       @options = options.dup
     end
 
     def query(&block)
-      query = Graphlient::Query.new do
+      query_str = Graphlient::Query.new do
         instance_eval(&block)
       end
-      response = post(query)
-      raise Graphlient::Errors::HTTP.new(response.message, response) unless response.is_a? Net::HTTPSuccess
-      parse(response.body)
-    end
-
-    def connection
-      @connection ||= Net::HTTP.new(uri.host, uri.port)
+      parsed_query = client.parse(query_str.to_s)
+      client.allow_dynamic_queries = true
+      client.query(parsed_query, context: @options)
+    rescue GraphQL::Client::Error => e
+      raise Graphlient::Errors::Client.new(e.message, e)
     end
 
-    def post(query)
-      request = Net::HTTP::Post.new(uri.request_uri)
-      request.body = { query: query.to_s }.to_json
-      options[:headers].each do |k, v|
-        request[k] = v
-      end
-      connection.request(request)
-    end
+    private
 
-    def parse(response)
-      JSON.parse(response, symbolize_names: true)
+    def client
+      return @client if @client
```

Comment:
Nitpick, `@client ||= begin` is, I believe, threadsafe in Ruby while this is not.